docs/src/main/asciidoc/writing-extensions.adoc

```diff
@@ -2025,16 +2025,103 @@ public final class MyExtProcessor {
+<2> Indicate that the `META-INF/my-descriptor.xml` resource should be copied into native image
 <3> Indicate that the "javax.xml.bind.Messages" resource bundle should be copied into native image.
 
+==== Service files
+
+If you are using `META-INF/services` files you need to register the files as resources so that your native image can find them,
+but you also need to register each listed class for reflection so they can be instantiated or inspected at run-time:
+
+[source,java]
+----
+public final class MyExtProcessor {
+
+    @BuildStep
+    void registerNativeImageReources(BuildProducer<NativeImageResourceBuildItem> resource,
+                                     BuildProducer<ReflectiveClassBuildItem> reflectionClasses) {
+        String service = "META-INF/services/" + io.quarkus.SomeService.class.getName();
+        
+        // register the service file so it is visible in native-image
+        resource.produce(new NativeImageResourceBuildItem(service));
+        
+        // register every listed implementation class so they can be inspected/instantiated 
+        // in native-image at run-time
+        Set<String> implementations = 
+            ServiceUtil.classNamesNamedIn(Thread.currentThread().getContextClassLoader(),
+                                          service);
+        reflectionClasses.produce(
+            new ReflectiveClassBuildItem(true, true, implementations.toArray(new String[0])));
+    }
+}
+----
+
+This is the easiest way to get your services running natively, but it's less efficient than scanning the implementation
+classes at build time and generating code that registers them at static-init time instead of relying on reflection.
+
+You can achieve that by adapting the previous build step to use a static-init recorder instead of registering 
+classes for reflection:
+
+[source,java]
+----
+public final class MyExtProcessor {
+
+    @BuildStep
+    @Record(ExecutionTime.STATIC_INIT)
+    void registerNativeImageReources(RecorderContext recorderContext, 
+                                     SomeServiceRecorder recorder) {
+        String service = "META-INF/services/" + io.quarkus.SomeService.class.getName();
+        
+        // read the implementation classes
+        Collection<Class<? extends io.quarkus.SomeService>> implementationClasses = new LinkedHashSet<>();
+        Set<String> implementations = ServiceUtil.classNamesNamedIn(Thread.currentThread().getContextClassLoader(),
+                                                                    service);
+        for(String implementation : implementations) {
+            implementationClasses.add((Class<? extends io.quarkus.SomeService>) 
+                recorderContext.classProxy(implementation));
+        }
+        
+        // produce a static-initializer with those classes
+        recorder.configure(implementationClasses);
+    }
+}
+
+@Recorder
```

Comment:
I think I would separate the two classes with a sentence making it clear that they are in 2 different contexts (deployment vs runtime artifact)

Reply:
OK, then I should make another PR, right?

Reply:
Yes, please. With the `backport?` tag too.